MemoryWidget: Add Symbols and Notes.

TryTwo opened the pull request:
Adds a note search box to memory widget
Adds a symbols/notes column to MemoryView
Adds option to hide all that in the memory widget views menu.

I'm not sure how interactable normal symbols should be.  Should they be mostly managed in CodeView?  I don't currently have them searchable and editable in the MemoryView, but they will be displayed if no Note is at that address.

![image](https://github.com/user-attachments/assets/eb9a47c5-57b8-4de1-8614-d2e3acb9d73a)

TryTwo commented:
Memory can edit and control notes the same as code widget. The thing I left out was normal symbol editing and searching from memory widget.  Are there some symbol maps that go to memory values (non-code)?

dreamsyntax commented:
> Memory can edit and control notes the same as code widget. The thing I left out was normal symbol editing and searching from memory widget.  Are there some symbol maps that apply to memory values (non-code)?

Yes, for example if you use IDA/Ghidra and automap memory strings. Dolphin shows those as grayed out text in the CodeWidget, but clicking does nothing.

dreamsyntax commented:
Ok, so I don't think this is that useful as a 'notes' utilization.
The idea is great though but should apply to .data,
like a map with
`.data section layout`, example:
https://github.com/ShadowTheHedgehogHacking/shadow-symbols-and-dmw/commit/d92b8dd448b40cdbc792e5f5fc288c58c14f18d5

TryTwo commented:
I refactored EditSymbolDialog to handle all the editing itself and added Data editing to memory widget.  It's a little laggy, so I need to add a custom function for updating the symbol column to MemoryView.

Changing to draft, because I probably need to PR the refactor separately, or else it'll be a pain to review.

TryTwo commented:
I moved the EditSymbolDialog refactor to be first, then the memory widget symbols second.  It could be reviewed now, but I could break things up more too.

TryTwo commented:
After looking it over one more time, I removed the EditSymbolDialog refactor. It's not too complicated without it, and I can PR it separately based on reviewer feedback.

dreamsyntax commented:
Is this considered done then? I can give it a once over

TryTwo commented:
Yes. I think it's done. Give it a go!

sepalani reviewed (changes requested):
That's the kind of feature I didn't know I needed, it's really useful (like its counterpart in the Code widget). I tested it a bit and beside these points, the code looks fine to me.

Here is an updated screenshot of what it currently looks like:
<img width="2560" height="1528" alt="image" src="https://github.com/user-attachments/assets/913b3b1c-62df-4bf6-82a3-83baabb2bb15" />

TryTwo commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -1059,6 +1106,76 @@ void MemoryViewWidget::OnCopyHex(u32 addr)
+  if (type == EditSymbolType::EditRegion)
+  {
+    // If symbol doesn't exist, it's safe to add a new region.
+    Common::Symbol* symbol = m_ppc_symbol_db.GetSymbolFromAddr(addr);
```
Yes. I think I can do const pointer const?

TryTwo commented:
Thanks for the review! I think the right side menu is getting a little crowded. Maybe I can figure something out for that in another PR.

sepalani reviewed (approved):
Tested and LGTM.
